File: backend/app/api/v1/endpoints/sla.py
"""
SmartComplaintHandler - SLA & Lifecycle Endpoints
Blueprint Reference: V1/M5/backend/05_sla_endpoints.md
Role: REST routes for status updates (/status), ticket resolution (/resolve), and active breach queries.
"""
from typing import List
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.schemas.sla import (
    StatusUpdateRequest,
    TicketResolveRequest,
    EscalationRequest,
    SLABreachResponse,
    TicketLifecycleResponse,
)
from app.services.ticket_service import (
    update_ticket_status,
    resolve_ticket,
    escalate_ticket,
    get_active_sla_breaches,
)
from app.services.lifecycle import (
    InvalidStateTransitionError,
    MissingResolutionNotesError,
    TerminalStateModificationError,
)

router = APIRouter()


@router.patch("/tickets/{ticket_id}/status", response_model=TicketLifecycleResponse)
def update_status(
    ticket_id: int,
    payload: StatusUpdateRequest,
    db: Session = Depends(get_db),
):
    """
    Advance or update ticket lifecycle status through formal transition validation.
    """
    try:
        ticket = update_ticket_status(
            db=db,
            ticket_id=ticket_id,
            new_status=payload.status.value,
            staff_notes=payload.notes,
            actor=payload.actor,
        )
    except (InvalidStateTransitionError, TerminalStateModificationError) as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        )

    if ticket is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Ticket with ID {ticket_id} not found.",
        )

    return ticket


@router.post("/tickets/{ticket_id}/resolve", response_model=TicketLifecycleResponse)
def resolve(
    ticket_id: int,
    payload: TicketResolveRequest,
    db: Session = Depends(get_db),
):
    """
    Formally close and resolve an incident, capturing mandatory repair notes.
    """
    try:
        ticket = resolve_ticket(
            db=db,
            ticket_id=ticket_id,
            resolution_notes=payload.resolution_notes,
            parts_replaced=payload.parts_replaced,
            technician=payload.technician_name,
        )
    except (InvalidStateTransitionError, MissingResolutionNotesError, TerminalStateModificationError) as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        )

    if ticket is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Ticket with ID {ticket_id} not found.",
        )

    return ticket


@router.post("/tickets/{ticket_id}/escalate", response_model=TicketLifecycleResponse)
def escalate(
    ticket_id: int,
    payload: EscalationRequest,
    db: Session = Depends(get_db),
):
    """
    Escalate a stagnant or high-risk ticket for supervisory oversight.
    """
    try:
        ticket = escalate_ticket(
            db=db,
            ticket_id=ticket_id,
            reason=payload.escalation_reason,
            supervisor_id=payload.supervisor_id,
        )
    except InvalidStateTransitionError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        )

    if ticket is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Ticket with ID {ticket_id} not found.",
        )

    return ticket


@router.get("/sla/breaches/active", response_model=List[SLABreachResponse])
def get_breaches(
    threshold_ratio: float = Query(0.20, ge=0.0, le=1.0, description="Approaching breach warning threshold ratio"),
    db: Session = Depends(get_db),
):
    """
    Query active tickets that have breached or are approaching breach of contractual SLA.
    """
    breaches = get_active_sla_breaches(db=db, threshold_ratio=threshold_ratio)
    return breaches

File: frontend/src/api/sla.js
/**
 * SmartComplaintHandler - SLA API Transport
 * Blueprint Reference: V1/M5/frontend/01_sla_api_client.md
 * Role: HTTP functions for status updates, ticket resolution, and breach monitoring.
 */
import apiClient from './client';

/**
 * Advance or modify ticket lifecycle status with optional transition remarks.
 *
 * @param {number} ticketId - Unique ticket primary key.
 * @param {string} newStatus - Target lifecycle state (e.g. 'IN_PROGRESS', 'ON_HOLD').
 * @param {string|null} notes - Optional transition comments or rationale.
 * @param {string} actor - Identity of staff member making the change.
 * @returns {Promise<Object>} Updated ticket lifecycle response entity.
 */
export async function updateTicketStatus(ticketId, newStatus, notes = null, actor = 'Staff') {
  if (!ticketId || typeof ticketId !== 'number' || ticketId <= 0) {
    throw new TypeError('A valid positive integer ticketId is required.');
  }
  if (!newStatus || typeof newStatus !== 'string') {
    throw new TypeError('A valid target status string is required.');
  }

  const payload = {
    status: newStatus.trim().toUpperCase(),
    notes: notes && notes.trim() ? notes.trim() : null,
    actor: actor || 'Staff'
  };

  const response = await apiClient.patch(`/tickets/${ticketId}/status`, payload);
  return response.data;
}

/**
 * Formally resolve an incident, enforcing substantive closure notes.
 *
 * @param {number} ticketId - Unique ticket primary key.
 * @param {string} resolutionNotes - Mandatory repair notes (min 10 characters).
 * @param {string|null} partsReplaced - Optional summary of replacement parts used.
 * @param {string|null} technicianName - Optional name of technician who repaired it.
 * @returns {Promise<Object>} Resolved ticket entity with resolution timestamp.
 */
export async function resolveTicket(ticketId, resolutionNotes, partsReplaced = null, technicianName = null) {
  if (!ticketId || typeof ticketId !== 'number' || ticketId <= 0) {
    throw new TypeError('A valid positive integer ticketId is required.');
  }
  if (!resolutionNotes || typeof resolutionNotes !== 'string' || resolutionNotes.trim().length < 10) {
    throw new Error('Resolution notes must contain at least 10 characters detailing the repair performed.');
  }

  const payload = {
    resolution_notes: resolutionNotes.trim(),
    parts_replaced: partsReplaced && partsReplaced.trim() ? partsReplaced.trim() : null,
    technician_name: technicianName && technicianName.trim() ? technicianName.trim() : null
  };

  const response = await apiClient.post(`/tickets/${ticketId}/resolve`, payload);
  return response.data;
}

/**
 * Escalate a stagnant or high-risk ticket for supervisory oversight.
 *
 * @param {number} ticketId - Unique ticket primary key.
 * @param {string} reason - Mandatory justification for escalation (min 5 characters).
 * @param {string|null} supervisorId - Optional supervisor identifier.
 * @returns {Promise<Object>} Escalated ticket entity.
 */
export async function escalateTicket(ticketId, reason, supervisorId = null) {
  if (!ticketId || typeof ticketId !== 'number' || ticketId <= 0) {
    throw new TypeError('A valid positive integer ticketId is required.');
  }
  if (!reason || typeof reason !== 'string' || reason.trim().length < 5) {
    throw new Error('Escalation reason must contain at least 5 characters.');
  }

  const payload = {
    escalation_reason: reason.trim(),
    supervisor_id: supervisorId && supervisorId.trim() ? supervisorId.trim() : null
  };

  const response = await apiClient.post(`/tickets/${ticketId}/escalate`, payload);
  return response.data;
}

/**
 * Query active tickets that are overdue or nearing deadline breach.
 *
 * @param {number} thresholdRatio - Ratio of SLA duration triggering warning (default: 0.20).
 * @returns {Promise<Array>} Array of breach summary objects.
 */
export async function fetchActiveBreaches(thresholdRatio = 0.20) {
  const response = await apiClient.get('/sla/breaches/active', {
    params: { threshold_ratio: thresholdRatio }
  });
  return response.data;
}
